lib/predicate/registry.js

```diff
@@ -0,0 +1,367 @@
+    bucket: eventInfoSpec.bucket,
+    key: /\w+/,
+    // TODO: improve field extraction
+    request(req) {
+        if (!(req instanceof stream.Readable)) {
+            return false;
+        }
+        const errs = validate({
+            headers: {},
+            namespace: '',
+            parsedHost(ip) {
+                if (!ip || typeof ip !== 'string') {
+                    return false;
+                }
+                const parts = ip.split('.');
+                if (parts.length !== 4) {
+                    return false;
+                }
+                return parts.every(p => {
+                    if (isNaN(p) || p > 255) {
+                        return false;
+                    }
+                    return true;
+                });
+            },
+            parsedContentLength: 0,
+            contentMD5: /^[0-9a-f]{32}$/i,
+        }, req);
+        if (errs.length) {
+            return false;
+        }
+        this.headers = req.headers;
+        this.region = req.namespace;
+        this.sourceIPAddress = req.parsedHost;
+        this.size = req.parsedContentLength;
+        this.etag = req.contentMD5;
+        this.contentType = req.headers['content-type'];
+        this.metadata = {};
+        Object.keys(req.headers).forEach(k => {
+            const m = metadataRe.exec(k);
+            if (!m) {
+                return;
+            }
+            this.metadata[m[1]] = req.headers[k];
+        });
+        return true;
+    },
+};
+
+function createKey(eventInfo) {
+    return `${eventInfo.eventName}|${eventInfo.bucket}`;
+}
+
+// TODO: improve field extraction
+function makePredicateEvent(runParams, buf) {
+    return {
+        Records: [{
+            eventVersion: '2.0',
+            eventSource: 'aws:s3',
+            awsRegion: runParams.region,
+            eventTime: new Date(),
+            eventName: runParams.eventName,
+            userIdentity: {
+                principalId: runParams.userId,
+            },
+            requestParameters: {
+                sourceIPAddress: runParams.sourceIPAddress,
+            },
+            responseElements: {
+                'x-amz-request-id': runParams.amzRequestId,
+                'x-amz-id-2': runParams.amzRequestId2,
+            },
+            s3: {
+                s3SchemaVersion: '1.0',
+                configurationId: '00000000-0000-0000-0000-000000000000',
+                bucket: {
+                    name: runParams.bucket,
+                    ownerIdentity: {
+                        principalId: runParams.bucketOwnerId,
+                    },
+                    arn: runParams.bucketArn,
+                },
+                object: {
+                    key: runParams.key,
+                    size: runParams.size,
+                    eTag: runParams.etag,
+                    contentType: runParams.contentType,
+                    metadata: runParams.metadata,
+                    data: buf,
```

Comment:
I was trying to keep the API simple, but your streaming suggestion sounds better